vendor/github.com/ethereum/go-ethereum/whisper/whisperv5/whisper.go

```diff
@@ -718,6 +735,33 @@ func (wh *Whisper) add(envelope *Envelope) (bool, error) {
 	return true, nil
 }
 
+func (w *Whisper) traceIncomingDelivery(isP2P bool, status message.Status, src NewMessage, env *Envelope, rec *ReceivedMessage, err error) {
+	w.traceDelivery(isP2P, message.IncomingMessage, status, src, env, rec, err)
+}
+
+func (w *Whisper) traceOutgoingDelivery(isP2P bool, status message.Status, src NewMessage, env *Envelope, rec *ReceivedMessage, err error) {
+	w.traceDelivery(isP2P, message.OutgoingMessage, status, src, env, rec, err)
+}
+
+func (w *Whisper) traceDelivery(isP2P bool, dir message.Direction, status message.Status, src NewMessage, env *Envelope, rec *ReceivedMessage, err error) {
+	if w.deliveryServer != nil {
+		var envelope Envelope
+
+		if env != nil {
+			envelope = *env
+		}
+
+		w.deliveryServer.SendState(MessageState{
```

Comment:
We know, that this function is only going to log message, but technically there is no guarantee. Also, on a huge load, logging could be a bottlneck (we use pretty expensive logging library), and we don't care about result here, so maybe worth start it as a short-lived goroutine? At least, it will guarantee that whisper code doesn't stuck because of misbihaved/overloaded logger.

Reply:
Hmmm, not bad idea. I think I have a perfect fit for this, will just change `LogDeliveryService` for this.

Reply:
hmmmm....using the goroutine version would make sense, but then naturally the developer must also take care to ensure the goroutine does not leak. It's tradeoff here.

Reply:
But naturally i think this should not cause to much harm.